File: src/main/java/tech/omnidigit/gitexporter/ui/CommitPreviewDialog.java
package tech.omnidigit.gitexporter.ui;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.vcs.changes.Change;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.components.JBScrollPane;
import git4idea.GitCommit;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.List;

/*
 * Copyright (c) 2025 Tony Ben
 */

public class CommitPreviewDialog extends DialogWrapper {

    private final Project project;

    private final JList<GitCommit> commitList;
    private final DefaultListModel<GitCommit> model;
    private final Map<GitCommit, Boolean> commitSelections;
    private final JCheckBox exportClassFiles;

    private final JTree fileTree;
    private final DefaultTreeModel treeModel;

    public CommitPreviewDialog(Project project) {
        super(project);

        this.project = project;
        model = new DefaultListModel<>();
        commitList = new JList<>(model);
        commitList.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);

        // 初始化文件树
        treeModel = new DefaultTreeModel(new DefaultMutableTreeNode("请选择一个提交"));
        fileTree = new JTree(treeModel);
        fileTree.setRootVisible(false);
        fileTree.setShowsRootHandles(true);

        // 创建复选框映射
        commitSelections = new HashMap<>();

        // 设置自定义渲染器
        commitList.setCellRenderer(new CommitListRenderer());

        commitList.addListSelectionListener(e -> {
            if (!e.getValueIsAdjusting()) {
                GitCommit selected = commitList.getSelectedValue();
                if (selected != null) {
                    updateFileTree(selected);
                }
            }
        });

        // 添加鼠标监听器处理选择
        commitList.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                int index = commitList.locationToIndex(e.getPoint());
                if (index >= 0) {
                    boolean flag = false;
                    Rectangle rect = commitList.getCellBounds(index, index);
                    if (e.getClickCount() == 1 && rect.contains(e.getPoint()) && e.getX() <= rect.x + 20) {
                        flag = true;
                    } else if (e.getClickCount() == 2 && rect.contains(e.getPoint()) && e.getX() > rect.x + 20) {
                        flag = true;
                    }
                    if (flag) {
                        GitCommit commit = model.getElementAt(index);
                        toggleSelection(commit);
                    }
                }
            }
        });

        // 添加导出选项
        exportClassFiles = new JCheckBox("同时导出对应的.class文件", true);

        setTitle("选择要导出的提交");
        init();
    }

    @Override
    protected JComponent createCenterPanel() {


        JSplitPane splitPane = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT);

        JPanel commitPanel = new JPanel(new BorderLayout());
        // 添加滚动面板
        commitPanel.add(new JBScrollPane(commitList), BorderLayout.CENTER);

        // 添加按钮面板
        JPanel buttonPanel = new JPanel();
        JButton selectAllButton = new JButton("全选");
        JButton deselectAllButton = new JButton("取消全选");
        JButton invertSelectionButton = new JButton("反选");

        selectAllButton.addActionListener(e -> selectAllCommits());
        deselectAllButton.addActionListener(e -> deselectAllCommits());
        invertSelectionButton.addActionListener(e -> toggleSelection());

        buttonPanel.add(selectAllButton);
        buttonPanel.add(deselectAllButton);
        buttonPanel.add(invertSelectionButton);

        commitPanel.add(buttonPanel, BorderLayout.NORTH);
        commitPanel.add(exportClassFiles, BorderLayout.SOUTH);

        // 文件树面板
        JPanel fileTreePanel = new JPanel(new BorderLayout());
        fileTreePanel.add(new JBScrollPane(fileTree), BorderLayout.CENTER);
        fileTreePanel.setBorder(BorderFactory.createTitledBorder("修改的文件"));
        fileTreePanel.setPreferredSize(new Dimension(400, 200));

        // 主布局
        splitPane.setLeftComponent(commitPanel);
        splitPane.setRightComponent(fileTreePanel);
        splitPane.setDividerLocation(0.5); // 初始分割比例为50%

        return splitPane;
    }

    private void updateFileTree(GitCommit commit) {
        DefaultMutableTreeNode root = new DefaultMutableTreeNode("修改的文件");
        Collection<VirtualFile> files = getFiles(commit);

        if (files != null && !files.isEmpty()) {
            // 获取项目基础路径
            String basePath = project.getBasePath();
            int basePathLength = basePath != null ? basePath.lastIndexOf("/") : 0;

            // 构建路径树
            Map<String, DefaultMutableTreeNode> nodeMap = new HashMap<>();
            nodeMap.put("", root);

            for (VirtualFile file : files) {
                String fullPath = file.getPath();
                // 从完整路径中去掉项目基础路径
                String relativePath = basePath != null && fullPath.startsWith(basePath)
                        ? fullPath.substring(basePathLength)
                        : fullPath;

                // 去掉开头的斜杠
                if (relativePath.startsWith("/")) {
                    relativePath = relativePath.substring(1);
                }

                String[] parts = relativePath.split("/");
                DefaultMutableTreeNode parent = root;
                StringBuilder currentPath = new StringBuilder();

                for (int i = 0; i < parts.length; i++) {
                    if (i > 0) currentPath.append("/");
                    currentPath.append(parts[i]);

                    String pathKey = currentPath.toString();
                    DefaultMutableTreeNode node = nodeMap.get(pathKey);

                    if (node == null) {
                        node = new DefaultMutableTreeNode(parts[i]);
                        parent.add(node);
                        nodeMap.put(pathKey, node);
                    }
                    parent = node;
                }
            }
        }

        treeModel.setRoot(root);
        expandAllNodes(fileTree, 0, fileTree.getRowCount());
    }

    private void expandAllNodes(JTree tree, int startingIndex, int rowCount) {
        for (int i = startingIndex; i < rowCount; ++i) {
            tree.expandRow(i);
        }

        if (tree.getRowCount() != rowCount) {
            expandAllNodes(tree, rowCount, tree.getRowCount());
        }
    }

    private void toggleSelection(GitCommit commit) {
        boolean selected = !commitSelections.getOrDefault(commit, false);
        commitSelections.put(commit, selected);
        commitList.repaint();
    }

    // 全选
    private void selectAllCommits() {
        commitSelections.forEach((commit, selected) -> {
            commitSelections.put(commit, true);
        });
        commitList.repaint();
    }

    // 取消全选
    private void deselectAllCommits() {
        commitSelections.forEach((commit, selected) -> {
            commitSelections.put(commit, false);
        });
        commitList.repaint();
    }

    // 反选
    private void toggleSelection() {
        commitSelections.forEach((commit, selected) -> {
            commitSelections.put(commit, !selected);
        });
        commitList.repaint();
    }

    public List<GitCommit> getSelectedCommits() {
        return commitSelections.entrySet().stream()
                .filter(Map.Entry::getValue)
                .map(Map.Entry::getKey)
                .toList();
    }

    public boolean shouldExportClassFiles() {
        return exportClassFiles.isSelected();
    }

    public void addCommits(List<GitCommit> commits) {
        commits.forEach(this::addCommit);
    }

    // 添加新的提交记录
    public void addCommit(GitCommit commit) {
        model.addElement(commit);
        commitSelections.put(commit, false);
    }

    // 删除指定的提交记录
    public void removeCommit(GitCommit commit) {
        model.removeElement(commit);
        commitSelections.remove(commit);
    }

    // 清空所有提交记录
    public void clearCommits() {
        model.clear();
        commitSelections.clear();
    }

    // 自定义列表渲染器
    private class CommitListRenderer extends JPanel implements ListCellRenderer<GitCommit> {
        private final JCheckBox checkBox = new JCheckBox();
        private final JLabel label = new JLabel();

        private GitCommit currentCommit;

        private final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss"); // 定义时间格式

        public CommitListRenderer() {
            setOpaque(true);
            setLayout(new BorderLayout());
            add(checkBox, BorderLayout.WEST);
            add(Box.createHorizontalStrut(10), BorderLayout.CENTER);
            add(label, BorderLayout.CENTER);
        }

        @Override
        public Component getListCellRendererComponent(JList<? extends GitCommit> list, GitCommit commit, int index,
                                                      boolean isSelected, boolean cellHasFocus) {
            currentCommit = commit;

            label.setText(String.format("%s - %s", dateFormat.format(commit.getAuthorTime()), commit.getSubject()));
            checkBox.setSelected(commitSelections.getOrDefault(commit, false));

            setBackground(isSelected ? list.getSelectionBackground() : list.getBackground());
            label.setForeground(isSelected ? list.getSelectionForeground() : list.getForeground());
            checkBox.setForeground(isSelected ? list.getSelectionForeground() : list.getForeground());

            return this;
        }
    }


    private Collection<VirtualFile> getFiles(GitCommit commit) {
        Collection<VirtualFile> files = commit.getChanges().stream()
                .map(Change::getVirtualFile)
                .filter(Objects::nonNull)
                .toList();
        return files;
    }
}

File: src/main/kotlin/tech/omnidigit/gitexporter/ui/CommitPreviewDialog.kt
package tech.omnidigit.gitexporter.ui

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.ui.components.JBScrollPane
import git4idea.GitCommit
import tech.omnidigit.gitexporter.resource.CommitPreviewBundle
import java.awt.BorderLayout
import java.awt.Component
import java.awt.Dimension
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.text.SimpleDateFormat
import javax.swing.*
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel

/*
 * Copyright (c) 2025 Tony Ben
 */

class CommitPreviewDialog(private val project: Project) : DialogWrapper(project) {

    private val model = DefaultListModel<GitCommit>()
    private val commitList = JList(model).apply {
        selectionMode = ListSelectionModel.SINGLE_SELECTION
        cellRenderer = CommitListRenderer()
    }

    private val treeModel = DefaultTreeModel(
        DefaultMutableTreeNode(CommitPreviewBundle.message("placeholder.select_commit"))
    )
    private val fileTree = JTree(treeModel).apply {
        isRootVisible = false
        showsRootHandles = true
    }

    private val commitSelections = mutableMapOf<GitCommit, Boolean>()
    private val exportClassFiles = JCheckBox(
        CommitPreviewBundle.message("checkbox.export_class"),
        true
    )
    private val dateFormat = SimpleDateFormat("yyyy/MM/dd HH:mm:ss")

    init {
        title = CommitPreviewBundle.message("dialog.title")
        init()

        commitList.addListSelectionListener { e ->
            if (!e.valueIsAdjusting) {
                commitList.selectedValue?.let { updateFileTree(it) }
            }
        }

        commitList.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                val index = commitList.locationToIndex(e.point)
                if (index >= 0) {
                    val rect = commitList.getCellBounds(index, index)
                    val shouldToggle = when {
                        e.clickCount == 1 && rect.contains(e.point) && e.x <= rect.x + 20 -> true
                        e.clickCount == 2 && rect.contains(e.point) && e.x > rect.x + 20 -> true
                        else -> false
                    }
                    if (shouldToggle) {
                        model.getElementAt(index)?.let { toggleSelection(it) }
                    }
                }
            }
        })
    }

    override fun createCenterPanel(): JComponent {
        return JSplitPane(JSplitPane.HORIZONTAL_SPLIT).apply {
            leftComponent = createCommitPanel()
            rightComponent = createFileTreePanel()
        }
    }

    private fun createCommitPanel(): JPanel {
        return JPanel(BorderLayout()).apply {
            add(JBScrollPane(commitList), BorderLayout.CENTER)
            add(createButtonPanel(), BorderLayout.NORTH)
            add(exportClassFiles, BorderLayout.SOUTH)
        }
    }

    private fun createButtonPanel(): JPanel {
        return JPanel().apply {
            add(JButton(CommitPreviewBundle.message("button.select_all")).apply {
                addActionListener { selectAllCommits() }
            })
            add(JButton(CommitPreviewBundle.message("button.deselect_all")).apply {
                addActionListener { deselectAllCommits() }
            })
            add(JButton(CommitPreviewBundle.message("button.invert")).apply {
                addActionListener { toggleSelection() }
            })
        }
    }

    private fun createFileTreePanel(): JPanel {
        return JPanel(BorderLayout()).apply {
            add(JBScrollPane(fileTree), BorderLayout.CENTER)
            border = BorderFactory.createTitledBorder(CommitPreviewBundle.message("tree.root"))
            preferredSize = Dimension(400, 200)
        }
    }

    private fun updateFileTree(commit: GitCommit) {
        val root = DefaultMutableTreeNode(CommitPreviewBundle.message("tree.root"))
        getFiles(commit)?.takeIf { it.isNotEmpty() }?.let { files ->
            val basePath = project.basePath
            val basePathLength = basePath?.lastIndexOf('/') ?: 0
            val nodeMap = mutableMapOf("" to root)

            files.forEach { file ->
                val fullPath = file.path
                val relativePath = basePath?.let { bp ->
                    if (fullPath.startsWith(bp)) fullPath.substring(basePathLength) else fullPath
                }?.removePrefix("/") ?: fullPath

                var parent = root
                val currentPath = StringBuilder()

                relativePath.split('/').forEach { part ->
                    if (currentPath.isNotEmpty()) currentPath.append('/')
                    currentPath.append(part)

                    val pathKey = currentPath.toString()
                    val node = nodeMap.getOrPut(pathKey) {
                        DefaultMutableTreeNode(part).also { parent.add(it) }
                    }
                    parent = node
                }
            }
        }

        treeModel.setRoot(root)
        expandAllNodes()
    }

    private fun expandAllNodes() {
        var row = 0
        while (row < fileTree.rowCount) {
            fileTree.expandRow(row++)
        }
    }

    private fun toggleSelection(commit: GitCommit) {
        commitSelections[commit] = !commitSelections.getOrDefault(commit, false)
        commitList.repaint()
    }

    private fun selectAllCommits() {
        commitSelections.keys.forEach { commitSelections[it] = true }
        commitList.repaint()
    }

    private fun deselectAllCommits() {
        commitSelections.keys.forEach { commitSelections[it] = false }
        commitList.repaint()
    }

    private fun toggleSelection() {
        commitSelections.keys.forEach { commitSelections[it] = !commitSelections[it]!! }
        commitList.repaint()
    }

    val selectedCommits: List<GitCommit>
        get() = commitSelections.filter { it.value }.keys.toList()

    val shouldExportClassFiles: Boolean
        get() = exportClassFiles.isSelected

    fun addCommits(commits: List<GitCommit>) {
        commits.forEach { addCommit(it) }
    }

    fun addCommit(commit: GitCommit) {
        model.addElement(commit)
        commitSelections[commit] = false
    }

    fun removeCommit(commit: GitCommit) {
        model.removeElement(commit)
        commitSelections.remove(commit)
    }

    fun clearCommits() {
        model.clear()
        commitSelections.clear()
    }

    private inner class CommitListRenderer : JPanel(BorderLayout()), ListCellRenderer<GitCommit> {
        private val checkBox = JCheckBox()
        private val label = JLabel()

        init {
            isOpaque = true
            add(checkBox, BorderLayout.WEST)
            add(Box.createHorizontalStrut(10), BorderLayout.CENTER)
            add(label, BorderLayout.CENTER)
        }

        override fun getListCellRendererComponent(
            list: JList<out GitCommit>,
            commit: GitCommit,
            index: Int,
            isSelected: Boolean,
            cellHasFocus: Boolean
        ): Component {
            label.text = "${dateFormat.format(commit.authorTime)} - ${commit.subject}"
            checkBox.isSelected = commitSelections.getOrDefault(commit, false)

            background = if (isSelected) list.selectionBackground else list.background
            label.foreground = if (isSelected) list.selectionForeground else list.foreground
            checkBox.foreground = label.foreground

            return this
        }
    }

    private fun getFiles(commit: GitCommit): Collection<VirtualFile>? {
        return commit.changes
            .mapNotNull { it.virtualFile }
            .takeIf { it.isNotEmpty() }
    }
}